### ISSUE TITLE
Item barcode required to display MFHD data and item status in Holdings tab

### ISSUE THREAD

**Issue by gibsonjc** [CARLI/vufind#242]
KNX discovered some items are not displaying items status.  Is it because the item doesn't have a barcode?

This KNX item has an "In Process" status but it does not have a barcode yet.
https://vufind3.carli.illinois.edu/vf-knx/Record/KNXdb.305459
![screen shot 2017-07-26 at 12 18 25 pm](https://user-images.githubusercontent.com/17598357/28634502-015abbb8-71fd-11e7-8276-88ddfcdbf40b.png)

But other KNX items that are "In Process" but have a barcode (and a call number) do show a status.
https://vufind3.carli.illinois.edu/vf-knx/Record/KNXdb.305291
https://vufind3.carli.illinois.edu/vf-knx/Record/KNXdb.302950
![screen shot 2017-07-26 at 12 20 42 pm](https://user-images.githubusercontent.com/17598357/28634469-e3cf4424-71fc-11e7-9037-d841a5ced254.png)

**Comment by dmcmpbll** [comment]
These KNX example may have changed; they have barcodes now:
https://vufind3.carli.illinois.edu/vf-knx/Record/KNXdb.305291
https://vufind3.carli.illinois.edu/vf-knx/Record/KNXdb.302950

-----------------------------------------------------------------------
From my initial tests it looks like the item needs to have a barcode for the status and copy number to display. The call number display does not seem to be barcode dependent.

ARU examples:

1)  https://vufind3.carli.illinois.edu/vf-aru/Record/ARUdb.25296 
     Library has 3 copies (1-3); none OPAC suppressed.
     All 3 have same NORMALIZED_CALL_NO and DISPLAY_CALL_NO
     Copy 1 has an item barcode; copies 2 & 3 do not.
2) https://vufind3.carli.illinois.edu/vf-aru/Record/ARUdb.449107
     Library has 1 copy with Copy number 0; not OPAC suppressed
     Item has no call number nor barcode.
3) https://vufind3.carli.illinois.edu/vf-aru/Record/ARUdb.129132
     Library has 2 copies (both copy number 1), neither OPAC suppressed.
     Both have same NORMALIZED_CALL_NO and DISPLAY_CALL_NO
     ITEM_ID 148589 has a barcode, ITEM_ID 148928 does not.

**Comment by dmcmpbll** [comment]
Was going to test the effects of adding a barcode, to double-check whether adding a barcode makes the status and copy number appear.

However, the titles look different between devel and prod, and between test and prod. (test and devel look the same). The basic data on the items as listed above appears unchanged between test/devel/prod.

129132 on Devel:
![129132devel](https://user-images.githubusercontent.com/20848151/28938559-07e88444-7854-11e7-8d36-2a1f8285d25a.PNG)

129132 on Prod:
![129132prod](https://user-images.githubusercontent.com/20848151/28938576-1515355e-7854-11e7-91e9-51d9edd98b8f.PNG)

**Comment by cedelis** [comment]
@dmcmpbll Can you set up an example on DEVEL? 

There are 2 Oracle SQL calls that are made to retrieve Holdings information for a particular bib. 

The first SQL statement is used to gather information for bibs which contain item records. The second one is used to gather those which have no item records. I have a feeling that the second SQL statement is not doing its job (in our environment, anyway).

Here's a reference to the comments and the code:

https://github.com/CARLI/vufind/blob/release-3.0/module/VuFind/src/VuFind/ILS/Driver/Voyager.php#L1110

Here's the cut-and-pasted comments from the above reference:

        // There are two possible queries we can use to obtain status information.
        // The first (and most common) obtains information from a combination of
        // items and holdings records.  The second (a rare case) obtains
        // information from the holdings record when no items are available.


I will cut and paste the two SQL statements currently being used by the above code, in case you'd like to run the queries yourself.

SQL for bibs w/ item records:

SELECT
   BIB_ITEM.BIB_ID,
   MFHD_ITEM.MFHD_ID,
   ITEM_BARCODE.ITEM_BARCODE,
   ITEM.ITEM_ID,
   ITEM.ON_RESERVE,
   ITEM.ITEM_SEQUENCE_NUMBER,
   ITEM.RECALLS_PLACED,
   ITEM.HOLDS_PLACED,
   ITEM_STATUS_TYPE.ITEM_STATUS_DESC as status,
   MFHD_DATA.RECORD_SEGMENT,
   MFHD_ITEM.ITEM_ENUM, NVL(LOCATION.LOCATION_DISPLAY_NAME, LOCATION.LOCATION_NAME) as location,
   ITEM.TEMP_LOCATION, ITEM.PERM_LOCATION, MFHD_MASTER.DISPLAY_CALL_NO as callnumber,
   to_char(CIRC_TRANSACTIONS.CURRENT_DUE_DATE, 'MM-DD-YY') as duedate,
   (SELECT TO_CHAR(MAX(CIRC_TRANS_ARCHIVE.DISCHARGE_DATE), 'MM-DD-YY HH24:MI') FROM ${sid}DB.CIRC_TRANS_ARCHIVE WHERE CI
RC_TRANS_ARCHIVE.ITEM_ID = ITEM.ITEM_ID) RETURNDATE,
   ITEM.ITEM_SEQUENCE_NUMBER,
   (SELECT SORT_GROUP_LOCATION.SEQUENCE_NUMBER FROM ${sid}DB.SORT_GROUP, ${sid}DB.SORT_GROUP_LOCATION WHERE SORT_GROUP.S
ORT_GROUP_DEFAULT = 'Y' AND SORT_GROUP_LOCATION.SORT_GROUP_ID = SORT_GROUP.SORT_GROUP_ID AND SORT_GROUP_LOCATION.LOCATIO
N_ID = ITEM.PERM_LOCATION) SORT_SEQ,
   ITEM.ITEM_TYPE_ID,
   ITEM.TEMP_ITEM_TYPE_ID FROM ${sid}DB.BIB_ITEM,
   ${sid}DB.ITEM,
   ${sid}DB.ITEM_STATUS_TYPE,
   ${sid}DB.ITEM_STATUS,
   ${sid}DB.LOCATION,
   ${sid}DB.MFHD_ITEM,
   ${sid}DB.MFHD_MASTER,
   ${sid}DB.MFHD_DATA,
   ${sid}DB.CIRC_TRANSACTIONS,
   ${sid}DB.ITEM_BARCODE WHERE BIB_ITEM.BIB_ID = ${bib_id} AND BIB_ITEM.ITEM_ID = ITEM.ITEM_ID AND ITEM.ITEM_ID = ITEM_S
TATUS.ITEM_ID AND ITEM_STATUS.ITEM_STATUS = ITEM_STATUS_TYPE.ITEM_STATUS_TYPE AND ITEM_BARCODE.ITEM_ID (+)= ITEM.ITEM_ID
 AND LOCATION.LOCATION_ID = ITEM.PERM_LOCATION AND CIRC_TRANSACTIONS.ITEM_ID (+)= ITEM.ITEM_ID AND MFHD_ITEM.ITEM_ID = I
TEM.ITEM_ID AND MFHD_MASTER.MFHD_ID = MFHD_ITEM.MFHD_ID AND MFHD_DATA.MFHD_ID = MFHD_ITEM.MFHD_ID AND MFHD_MASTER.SUPPRE
SS_IN_OPAC='N' ORDER BY ITEM.ITEM_SEQUENCE_NUMBER,
   MFHD_DATA.MFHD_ID,
   MFHD_DATA.SEQNUM

SQL for bibs w/out item records:

SELECT
   null as ITEM_BARCODE,
   null as ITEM_ID,
   MFHD_DATA.RECORD_SEGMENT,
   null as ITEM_ENUM,
   'N' as ON_RESERVE,
   1 as ITEM_SEQUENCE_NUMBER,
   'No information available' as status,
   NVL(LOCATION.LOCATION_DISPLAY_NAME, LOCATION.LOCATION_NAME) as location,
   MFHD_MASTER.DISPLAY_CALL_NO as callnumber,
   BIB_MFHD.BIB_ID,
   MFHD_MASTER.MFHD_ID,
   null as duedate,
   0 AS TEMP_LOCATION,
   0 as PERM_LOCATION,
   0 as ITEM_SEQUENCE_NUMBER,
   (SELECT SORT_GROUP_LOCATION.SEQUENCE_NUMBER FROM ${sid}DB.SORT_GROUP, ${sid}DB.SORT_GROUP_LOCATION WHERE SORT_GROUP.S
ORT_GROUP_DEFAULT = 'Y' AND SORT_GROUP_LOCATION.SORT_GROUP_ID = SORT_GROUP.SORT_GROUP_ID AND SORT_GROUP_LOCATION.LOCATIO
N_ID = LOCATION.LOCATION_ID) SORT_SEQ FROM ${sid}DB.BIB_MFHD,
   ${sid}DB.LOCATION,
   ${sid}DB.MFHD_MASTER,
   ${sid}DB.MFHD_DATA WHERE BIB_MFHD.BIB_ID = ${bib_id} AND LOCATION.LOCATION_ID = MFHD_MASTER.LOCATION_ID AND MFHD_MAST
ER.MFHD_ID = BIB_MFHD.MFHD_ID AND MFHD_DATA.MFHD_ID = BIB_MFHD.MFHD_ID AND MFHD_MASTER.SUPPRESS_IN_OPAC='N' AND NOT EXIS
TS (SELECT MFHD_ID FROM ${sid}DB.MFHD_ITEM WHERE MFHD_ITEM.MFHD_ID=MFHD_MASTER.MFHD_ID) ORDER BY MFHD_DATA.MFHD_ID,
   MFHD_DATA.SEQNUM

You will obviously need to supply the values for ${sid} and ${bib_id} above.

**Comment by dmcmpbll** [comment]
on ARU-devel, testing with https://vufind3-devel.carli.illinois.edu/vf-aru/Record/ARUdb.144810 
![before](https://user-images.githubusercontent.com/20848151/28945085-a5b15528-786a-11e7-90d5-04b4ec5c6c1f.PNG)
     1 item
     not suppressed
     Call Number of Lloyd
     copy 0
     no barcode
     not on reserves

1) Adjust Copy number to "1" in Circ
--No change in appearance from starting photo.

2) Add item barcode of "TEST" in Circ. 
Status now displays; copy number displays incorrectly as zero.
![afterbarcodeaftercopyupdatedto1](https://user-images.githubusercontent.com/20848151/28945290-840b7560-786b-11e7-9650-ab55cbe47ce2.PNG)

**Comment by cedelis** [comment]
Thanks, @dmcmpbll 

This seems to confirm the lack of barcode results in a missing display.

Can you please set up another example of a missing item barcode and leave it in that state?

Also, re: the copy number discrepancy, there is already an open issue: https://github.com/CARLI/vufind/issues/169

**Comment by dmcmpbll** [comment]
This ARU item does not have a barcode: https://vufind3-devel.carli.illinois.edu/vf-aru/Record/ARUdb.407433 
It is not suppressed, has a copy number of 1, and only one item on the bib.

The BIB_ID is 407433
MFHD_ID is 417349
ITEM_ID is 165650

**Comment by gibsonjc** [comment]
Not only Item Status, but also the call number and copy number and MFHD 853/863 or 866 holdings and notes.  This bib has a MFHD and an Item, but the Item does not have a barcode:
https://i-share.carli.illinois.edu/vf-aru/Record/ARUdb.407433

<img width="883" alt="screen shot 2018-04-06 at 10 16 08 am" src="https://user-images.githubusercontent.com/17598357/38429305-9f5813b2-3983-11e8-85ad-0e5e56fbac6c.png">

![screen shot 2018-04-06 at 10 16 24 am](https://user-images.githubusercontent.com/17598357/38429317-a42471ec-3983-11e8-9db5-1a5dcbcea3cd.png)

**Comment by gibsonjc** [comment]
GSU - I have a holdings record that I can't figure out why it is showing that we don't have any holdings available in New VuFind, but is fine in 0.6.

Replication Steps in New VuFind:
Do a title search for Hispania
Receive two records (1 print and 1 microform)
Click on the first result record and see holdings information
Click on the second result record and see "The catalog cannot display location or copy information at this time. Please ask a library staff member for assistance."

RT 98605

**Comment by gibsonjc** [comment]
GSU BIBS in question are:
- 85695 with MFHD 87247. Three Items attached but only one has a barcode (that one displays in the holdings area; the other two do not)
- 736228 with MFHD 828558. One Item attached but it is without a barcode (nothing displays in holdings area).

A MFHD must have at least one Item attached with a barcode in order for the MFHD data (call number, copy number, notes, holdings fields) to display in the Holdings tab area.    
--Item 101161 has a barcode so that's why <https://i-share.carli.illinois.edu/vf-gsu/Record/GSUdb.85695> displays the MFHD data and this Item in the Holdings tab.  
--<https://i-share.carli.illinois.edu/vf-gsu/Record/GSUdb.736228> does not have an Item with a barcode so no MFHD data will display.

A barcode must be in the Item record in order for that Item's information, status, and request link to display.  Items 379719 and 379720 (Bib 85695) and Item 421804 (Bib 736228) do not have barcodes.

**Comment by gibsonjc** [comment]
See email thread with Subject "RE: SCJ and VuFind4 all" on 4/6/2018 for Gordon's suggested SQL?

**Comment by gibsonjc** [comment]
@cedelis I also wanted to point out that there are two levels of problem reported here on this single Issue.  They may be related or not once you dig into it.  If we need to spin off a separate Issue, we can.

1. The data from a Voyager MFHD (Location, Call Number, Copy Number, any notes) is not displaying if there is not an Item records with a barcode attached to it.  Example: https://i-share-test.carli.illinois.edu/vf-aru/Record/ARUdb.407433

BIB 407433
MFHD 417349
Item NONE, so no MFHD data displays at all

What we want it to do: the MFHD data (Location, Call Number, Copy Number, any notes) will always display in the Holdings tab, even if there are zero Items attached to it, and if there are Items, whether or not any Items have barcodes.  Like VuFind 0.6 does: https://vufind.carli.illinois.edu/vf-aru/Record/aru_407433


2.  In a situation where there is a MFHD with more than one Item, if at least one Item has a barcode, that MFHD data will display, but Item information will only display for those items which have barcodes.  The Item information is not displaying if the Item does not have a barcode. 

What we want it to do: The Item information (copy number and data) will display even if the Item does not have a barcode.  

Example: https://i-share-test.carli.illinois.edu/vf-gsu/Record/GSUdb.85695

BIB 85695
MFHD 87247
Item 1 - ID 379720 - no barcode, no display of Item data "v.65-83 1982-2000 (MICROFICHE)"
Item 2 - ID 379719 - no barcode, no display of Item data "v.38-64 1955-1981"
Item 3 - ID 101161 - barcode! Item data displays as Copy 1 (v.1-37 1918-1954 (MICROFILM))

<img width="922" alt="screen shot 2018-04-12 at 8 45 19 pm" src="https://user-images.githubusercontent.com/17598357/38712606-74656efc-3e92-11e8-9682-2cc6e5d06290.png">

![screen shot 2018-04-12 at 8 45 56 pm](https://user-images.githubusercontent.com/17598357/38712625-8e5d42d0-3e92-11e8-8ca5-4537ace55c88.png)

**Comment by gibsonjc** [comment]
NBY - Our holdings records also include a url for our users to request an item for use in the building. While we hope to replace those urls,  for now we can't promote the New VuFind for our staff and users.
Examples:
https://i-share.carli.illinois.edu/vf-nby/Record/NBYdb.263203
https://i-share.carli.illinois.edu/vf-nby/Record/NBYdb.772883

RT 98774

**Comment by gibsonjc** [comment]
To sum up what we want:

The MFHD record data (852 field Location, Call Number, Copy Number, notes, and all 85X/86X data) should always display in the Holdings tab.  It should not matter if there are any Items attached to the MFHD or if those Items have barcodes or not.

All Item records' data should always display, even if the Item does not have a barcode in it.  I can spin this off into a separate Issue, if needed.

**Comment by gibsonjc** [comment]
IIT - RT 99170

**Comment by dmcmpbll** [comment]
Would it be helpful if I paste in an example of server SQL for the "SQL for bibs w/ item records" portion that accounts for items without barcodes? 
Looking at the SQL they use is complicated because it doesn't contain normal table joins. <-- edit I see that they're joining in the criteria instead (when reviewing in Access). Will try to write with that method.

**Comment by nmswanson** [comment]
UIU's example with no barcode- 

https://i-share.carli.illinois.edu/vf-uiu/Record/UIUdb.7863500#description

Patrons seem to have no way to request this item in VuFind 4.1

Compare to VuFind 0.6: request option present:

https://vufind.carli.illinois.edu/vf-uiu/Record/uiu_7863500/Request

One more note about this record. This title is part of the Purchase on Demand collection of records we have loaded into the catalog. We will have thousands of POD records like this and all without barcodes in the item records. And, I just learned about another group of records this is impacting. For newly received materials we have process set up where library patrons can request the items not cataloged yet but requesting will have us do rush cataloging to process quickly and get item to them. All of these items don't have barcodes yet and with the VuFind, without the barcode the request action is not showing as option to the user. 

Yes, we would also put this as a high priority to fix or we start to work through other options to address on our end (through some testing it looks like any presence of numbers or text there allows the services to display. Adding a fake barcode does then allow the Holding Record information to show. That would be a lot of work to go in and add that fake barcode though). 

CARLI has recommended not adding non-standard fake barcodes for this system display issue as it would require added work now and future cleanup later with a resolution to this priority issue

RT99715
